client/src/cmdhf15.c

```diff
@@ -1098,6 +1098,133 @@ static int CmdHF15ELoad(const char *Cmd) {
     return PM3_SUCCESS;
 }
 
+static int CmdHF15ESave(const char *Cmd) {
+    CLIParserContext *ctx;
+    CLIParserInit(&ctx, "hf 15 esave",
+                  "Save emulator memory into three files (BIN/EML/JSON) ",
+                  "hf 15 esave -f hf-15-01020304"
+                  "hf 15 esave -b 8 -c 42 -f hf-15-01020304"
+                 );
+    void *argtable[] = {
+        arg_param_begin,
+        arg_str1("f", "file", "<fn>", "filename of dump"),
+        arg_int0("b", "blocksize", "<dec>", "block size, defaults to 4"),
+        arg_int0("c", "count", "<dec>", "number of blocks to export, defaults to all"),
+        arg_param_end
+    };
+    CLIExecWithReturn(ctx, Cmd, argtable, true);
+
+    int fnlen = 0;
+    char filename[FILE_PATH_SIZE];
+    CLIParamStrToBuf(arg_get_str(ctx, 1), (uint8_t *)filename, FILE_PATH_SIZE, &fnlen);
+    int blocksize = arg_get_int_def(ctx, 2, 4);
+    int count = arg_get_int_def(ctx, 3, -1);
+    CLIParserFree(ctx);
+
+    int bytes = CARD_MEMORY_SIZE;
+    if (count > 0 && count * blocksize <= bytes) {
+        bytes = count * blocksize;
+    }
+
+    // reserve memory
+    uint8_t *dump = calloc(bytes, sizeof(uint8_t));
+    if (dump == NULL) {
+        PrintAndLogEx(WARNING, "Fail, cannot allocate memory");
+        return PM3_EMALLOC;
+    }
+
+    PrintAndLogEx(INFO, "Downloading %u bytes from emulator memory", bytes);
+    if (!GetFromDevice(BIG_BUF_EML, dump, bytes, 0, NULL, 0, NULL, 2500, false)) {
+        PrintAndLogEx(WARNING, "Fail, transfer from device time-out");
+        free(dump);
+        return PM3_ETIMEOUT;
+    }
+
+    saveFile(filename, ".bin", dump, bytes);
+    saveFileEML(filename, dump, bytes, blocksize);
+    saveFileJSON(filename, jsf15, dump, bytes, NULL);
+    free(dump);
+    return PM3_SUCCESS;
+}
+
+static void print_hrule(int blocksize) {
+    PrintAndLogEx(INFO, "-----+" NOLF);
+    for (int i = 0; i < blocksize; i++) {
+        PrintAndLogEx(NORMAL, "---" NOLF);
+    }
+    PrintAndLogEx(NORMAL, "-+-" NOLF);
+    for (int i = 0; i < blocksize; i++) {
+        PrintAndLogEx(NORMAL, "-" NOLF);
+    }
+    PrintAndLogEx(NORMAL, "");
+}
+
+static void print_blocks_15693(uint8_t *data, uint16_t bytes, int blocksize) {
+    int blocks = bytes / blocksize;
+    PrintAndLogEx(NORMAL, "");
+    print_hrule(blocksize);
+    PrintAndLogEx(INFO, " blk | data  " NOLF);
+    for (int i = 2; i < blocksize; i++) {
+        PrintAndLogEx(NORMAL, "   " NOLF);
+    }
```

Comment:
There should be a string formatter option to specify number of chars to be added. 
It will simplify this whole "dynamic create string headers based on block size" a bit

Reply:
Learned something new about format specifiers there. It indeed looks much better now.